rfcs/0051-emit-on-close-stream-executor.md

```diff
@@ -0,0 +1,158 @@
+
+And then we will see why and how it works in different situations.
+
+## Executors
+
+### Sort
+
+SortExecutor just has a executor which wraps a SortBuffer which can transform any stream into a append only stream ordered by a watermark column.
+
+- when stream chunk arrive
+  - write the chunk into stateTable
+  - write the chunk into SortBufferCache
+- when watermark arrive
+  - pass watermark into SortBufferCache
+- when barrier arrive
+  - drain complete rows from SortBufferCache, for each rows
+    - emit the row downstream
+    - delete the row in stateTable
+
+### SortAgg (with batch only agg calls)
+
+Under "EMIT ON WINDOW CLOSE" semantics, if there is watermark in group key, we can use SortExecutor for the input stream, and then we get a sorted stream which can be processed by "BatchSortAgg". Then, we can support all the aggregators which can run in batch mode.
+
+The SortAgg supports more kinds of aggregators under "EMIT ON WINDOW CLOSE", but should materialized all the input data, which could be worse than the normal streamGroupAgg. So we have another EMIT ON WINDOW CLOSE GroupAgg implementation as described below. I think the SortAgg is just to solve some aggregation functions which does not have a streaming version(hard to implement or with high cost)
+
+### GroupAgg
+
+If there is watermark in group key, the EMIT ON WINDOW CLOSE GroupAgg calculate the aggregation with the same logic with streamGroupAgg, But it only emit the complete part of the agg result. A simple idea is add a sortExecutor after a normal streamGroupAgg and we can find that the streamGroupAgg's result table and sortExecutor's stateTable exactly have the same schema and meaning.
+
+So the EMIT ON WINDOW CLOSE streamGroupAgg need only add a SortBufferCache on the agg's result table.
+
+- when stream chunk arrive
+  - do the same with normal streamGroupAgg
+- when watermark arrive
+  - pass watermark into SortBufferCache
+- when barrier arrive
+  - do the same with normal streamGroupAgg but not emit to down stream, just get the changes
+    - write the changes into SortBufferCache
+    - write the changes into result table
+    - (the LRU cache should has been changed by the agg's logic)
+  - drain complete rows from SortBufferCache, for each rows
+    - emit the row downstream
+    - delete the row in result table
+  - emit watermark with the last emitted row's watermark
+
+### OverAgg
+
+[RFC: Over Window on watermark](https://github.com/risingwavelabs/rfcs/pull/8)
+We still need more detailed design here, but it is clear that the row's complete and row's emit can not always happens at the same time.
+
+### EqJoin
+
+It likes window join in Flink.
+When there are watermarks in the first join key of the both two sides. We can use two SortBuffers(table and cache) to implement a sortMergeJoin. SortBuffer’s `peek` method is used here.
+
+### IntervalJoin
```

Comment:
Same thing here. Let me read the docs of Flink again to see whether there are better ideas...

Reply:
See https://github.com/risingwavelabs/rfcs/pull/32#discussion_r1059266190

In short, Flink puts join key before time column, rather than 2 copies of data. I think we may start with it first so that hopefully it can reuse the source code of the normal version of interval join operator.